File: swift-api/Sources/App/Repositories/TodoRepository.swift
import Foundation
import Logging
import PostgresNIO

protocol TodoRepository: Sendable {
    func create(_ todo: Todo) async throws -> Todo
    func findAll(userId: UUID) async throws -> [Todo]
    func findById(_ id: UUID, userId: UUID) async throws -> Todo?
    func update(_ todo: Todo) async throws -> Todo
    func delete(_ id: UUID, userId: UUID) async throws -> Bool
    func deleteAll(userId: UUID) async throws -> Int
}

struct TodoPostgresRepository: TodoRepository {
    let client: PostgresClient
    private let logger = Logger(label: "TodoRepository")

    func create(_ todo: Todo) async throws -> Todo {
        let rows = try await client.query(
            """
            INSERT INTO todos (id, user_id, title, "order", completed, url)
            VALUES (\(todo.id), \(todo.userId), \(todo.title), \(todo.order), \(todo.completed), \(todo.url))
            RETURNING id, user_id, title, "order", completed, url, created_at, updated_at
            """,
            logger: logger
        )

        for try await (id, uId, title, order, completed, url, createdAt, updatedAt) in rows.decode(
            (UUID, UUID, String, Int?, Bool, String?, Date, Date).self,
            context: .default
        ) {
            return Todo(
                id: id,
                userId: uId,
                title: title,
                order: order,
                completed: completed,
                url: url,
                createdAt: createdAt,
                updatedAt: updatedAt
            )
        }
        return todo
    }

    func findAll(userId: UUID) async throws -> [Todo] {
        let rows = try await client.query(
            """
            SELECT id, user_id, title, "order", completed, url, created_at, updated_at
            FROM todos
            WHERE user_id = \(userId)
            ORDER BY "order" NULLS LAST, created_at DESC
            """,
            logger: logger
        )

        var todos: [Todo] = []
        todos.reserveCapacity(16)
        for try await (id, uId, title, order, completed, url, createdAt, updatedAt) in rows.decode(
            (UUID, UUID, String, Int?, Bool, String?, Date, Date).self,
            context: .default
        ) {
            todos.append(
                Todo(
                    id: id,
                    userId: uId,
                    title: title,
                    order: order,
                    completed: completed,
                    url: url,
                    createdAt: createdAt,
                    updatedAt: updatedAt
                ))
        }
        return todos
    }

    func findById(_ id: UUID, userId: UUID) async throws -> Todo? {
        let rows = try await client.query(
            """
            SELECT id, user_id, title, "order", completed, url, created_at, updated_at
            FROM todos
            WHERE id = \(id) AND user_id = \(userId)
            """,
            logger: logger
        )

        for try await (id, uId, title, order, completed, url, createdAt, updatedAt) in rows.decode(
            (UUID, UUID, String, Int?, Bool, String?, Date, Date).self,
            context: .default
        ) {
            return Todo(
                id: id,
                userId: uId,
                title: title,
                order: order,
                completed: completed,
                url: url,
                createdAt: createdAt,
                updatedAt: updatedAt
            )
        }
        return nil
    }

    func update(_ todo: Todo) async throws -> Todo {
        let rows = try await client.query(
            """
            UPDATE todos
            SET title = \(todo.title),
                "order" = \(todo.order),
                completed = \(todo.completed),
                url = \(todo.url),
                updated_at = CURRENT_TIMESTAMP
            WHERE id = \(todo.id) AND user_id = \(todo.userId)
            RETURNING id, user_id, title, "order", completed, url, created_at, updated_at
            """,
            logger: logger
        )

        for try await (id, uId, title, order, completed, url, createdAt, updatedAt) in rows.decode(
            (UUID, UUID, String, Int?, Bool, String?, Date, Date).self,
            context: .default
        ) {
            return Todo(
                id: id,
                userId: uId,
                title: title,
                order: order,
                completed: completed,
                url: url,
                createdAt: createdAt,
                updatedAt: updatedAt
            )
        }
        return todo
    }

    func delete(_ id: UUID, userId: UUID) async throws -> Bool {
        try await client.query(
            """
            DELETE FROM todos
            WHERE id = \(id) AND user_id = \(userId)
            """,
            logger: logger
        )
        // If query executed without error, consider it successful
        return true
    }

    func deleteAll(userId: UUID) async throws -> Int {
        try await client.query(
            """
            DELETE FROM todos
            WHERE user_id = \(userId)
            """,
            logger: logger
        )
        return 0  // PostgresNIO doesn't return affected rows count easily
    }
}

File: go-api/internal/repository/todo_repository.go
package repository

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"todos-api/internal/models"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

var ErrTodoNotFound = errors.New("todo not found")

type TodoRepository struct {
	db *pgxpool.Pool
}

func NewTodoRepository(db *pgxpool.Pool) *TodoRepository {
	return &TodoRepository{db: db}
}

func (r *TodoRepository) Create(ctx context.Context, todo *models.Todo) error {
	todo.ID = uuid.New()
	todo.CreatedAt = time.Now()
	todo.UpdatedAt = time.Now()

	_, err := r.db.Exec(ctx,
		`INSERT INTO todos (id, user_id, title, "order", completed, url, created_at, updated_at)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
		todo.ID, todo.UserID, todo.Title, todo.Order, todo.Completed, todo.URL, todo.CreatedAt, todo.UpdatedAt,
	)
	return err
}

func (r *TodoRepository) FindAll(ctx context.Context, userID uuid.UUID) ([]models.Todo, error) {
	rows, err := r.db.Query(ctx,
		`SELECT id, user_id, title, "order", completed, url, created_at, updated_at
		 FROM todos WHERE user_id = $1
		 ORDER BY "order" NULLS LAST, created_at DESC`,
		userID,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	todos := make([]models.Todo, 0, 16)
	for rows.Next() {
		var todo models.Todo
		err := rows.Scan(&todo.ID, &todo.UserID, &todo.Title, &todo.Order, &todo.Completed, &todo.URL, &todo.CreatedAt, &todo.UpdatedAt)
		if err != nil {
			return nil, err
		}
		todos = append(todos, todo)
	}

	return todos, rows.Err()
}

func (r *TodoRepository) FindByID(ctx context.Context, id, userID uuid.UUID) (*models.Todo, error) {
	todo := &models.Todo{}
	err := r.db.QueryRow(ctx,
		`SELECT id, user_id, title, "order", completed, url, created_at, updated_at
		 FROM todos WHERE id = $1 AND user_id = $2`,
		id, userID,
	).Scan(&todo.ID, &todo.UserID, &todo.Title, &todo.Order, &todo.Completed, &todo.URL, &todo.CreatedAt, &todo.UpdatedAt)

	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrTodoNotFound
	}
	if err != nil {
		return nil, err
	}
	return todo, nil
}

func (r *TodoRepository) Update(ctx context.Context, todo *models.Todo) error {
	todo.UpdatedAt = time.Now()

	result, err := r.db.Exec(ctx,
		`UPDATE todos SET title = $1, "order" = $2, completed = $3, url = $4, updated_at = $5
		 WHERE id = $6 AND user_id = $7`,
		todo.Title, todo.Order, todo.Completed, todo.URL, todo.UpdatedAt, todo.ID, todo.UserID,
	)
	if err != nil {
		return err
	}

	if result.RowsAffected() == 0 {
		return ErrTodoNotFound
	}
	return nil
}

func (r *TodoRepository) UpdateFields(ctx context.Context, id, userID uuid.UUID, req *models.UpdateTodoRequest) (*models.Todo, error) {
	setClauses := []string{`updated_at = $1`}
	args := []interface{}{time.Now()}
	argIndex := 2

	if req.Title != nil {
		setClauses = append(setClauses, fmt.Sprintf(`title = $%d`, argIndex))
		args = append(args, *req.Title)
		argIndex++
	}
	if req.Order != nil {
		setClauses = append(setClauses, fmt.Sprintf(`"order" = $%d`, argIndex))
		args = append(args, *req.Order)
		argIndex++
	}
	if req.Completed != nil {
		setClauses = append(setClauses, fmt.Sprintf(`completed = $%d`, argIndex))
		args = append(args, *req.Completed)
		argIndex++
	}

	args = append(args, id, userID)
	query := fmt.Sprintf(
		`UPDATE todos SET %s WHERE id = $%d AND user_id = $%d
		 RETURNING id, user_id, title, "order", completed, url, created_at, updated_at`,
		strings.Join(setClauses, ", "), argIndex, argIndex+1,
	)

	todo := &models.Todo{}
	err := r.db.QueryRow(ctx, query, args...).Scan(
		&todo.ID, &todo.UserID, &todo.Title, &todo.Order, &todo.Completed,
		&todo.URL, &todo.CreatedAt, &todo.UpdatedAt,
	)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrTodoNotFound
	}
	if err != nil {
		return nil, err
	}
	return todo, nil
}

func (r *TodoRepository) Delete(ctx context.Context, id, userID uuid.UUID) error {
	result, err := r.db.Exec(ctx,
		`DELETE FROM todos WHERE id = $1 AND user_id = $2`,
		id, userID,
	)
	if err != nil {
		return err
	}

	if result.RowsAffected() == 0 {
		return ErrTodoNotFound
	}
	return nil
}

func (r *TodoRepository) DeleteAll(ctx context.Context, userID uuid.UUID) (int64, error) {
	result, err := r.db.Exec(ctx,
		`DELETE FROM todos WHERE user_id = $1`,
		userID,
	)
	if err != nil {
		return 0, err
	}
	return result.RowsAffected(), nil
}
